src/visor/blockdb/unspent.go

```diff
@@ -0,0 +1,297 @@
+	if err := tx.DeleteBucket(up.pool.Name); err != nil {
+		return err
+	}
+
+	_, err := tx.CreateBucket(up.pool.Name)
+	if err != nil {
+		return err
+	}
+
+	if err := tx.DeleteBucket(up.meta.Name); err != nil {
+		return err
+	}
+
+	_, err = tx.CreateBucket(up.meta.Name)
+	return err
+}
+
+func (up *UnspentPool) addWithTx(tx *bolt.Tx, ux coin.UxOut) (err error) {
+	// all updates will rollback if return is not nil
+	// in case of unexpected panic, we must catch it and return error
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("unspent pool add uxout failed: %v", err)
+		}
+	}()
+
+	// check if the uxout does exist in the pool
+	h := ux.Hash()
+	if up.Contains(h) {
+		return fmt.Errorf("attemps to insert uxout:%v twice into the unspent pool", h.Hex())
+	}
+
+	meta := unspentMeta{tx.Bucket(up.meta.Name)}
+	xorhash, err := meta.getXorHash()
+	if err != nil {
+		return err
+	}
+
+	xorhash = xorhash.Xor(ux.SnapshotHash())
+	if err := meta.setXorHash(xorhash); err != nil {
+		return err
+	}
+
+	return uxOut{tx.Bucket(up.pool.Name)}.set(h, ux)
+}
+
+// GetArray returns UxOut by given hash array, will return error when
+// if any of the hashes is not exist.
+func (up *UnspentPool) GetArray(hashes []cipher.SHA256) (coin.UxArray, error) {
+	uxs := make(coin.UxArray, 0, len(hashes))
+	for i := range hashes {
+		ux, ok, err := up.Get(hashes[i])
+		if err != nil {
+			return nil, err
+		}
+
+		if !ok {
+			return nil, fmt.Errorf("unspent output of %s does not exist", hashes[i].Hex())
+		}
+
+		uxs = append(uxs, ux)
+	}
+	return uxs, nil
+}
+
+// Get returns the uxout value of give hash
+func (up *UnspentPool) Get(h cipher.SHA256) (coin.UxOut, bool, error) {
+	if v := up.pool.Get(h[:]); v != nil {
+		var ux coin.UxOut
+		if err := encoder.DeserializeRaw(v, &ux); err != nil {
+			return coin.UxOut{}, false, fmt.Errorf("get uxout from pool failed: %v", err)
+		}
+		return ux, true, nil
+	}
+
+	return coin.UxOut{}, false, nil
+}
+
+// GetAll returns Pool as an array. Note: they are not in any particular order.
+func (up *UnspentPool) GetAll() (coin.UxArray, error) {
+	arr := make(coin.UxArray, 0, up.pool.Len())
+	if err := up.pool.ForEach(func(k, v []byte) error {
+		var uxout coin.UxOut
+		if err := encoder.DeserializeRaw(v, &uxout); err != nil {
+			return err
+		}
+
+		arr = append(arr, uxout)
+		return nil
+	}); err != nil {
+		return coin.UxArray{}, err
+	}
+
+	return arr, nil
+}
+
+// delete delete unspent of given hashes
+func (up *UnspentPool) deleteWithTx(tx *bolt.Tx, hashes []cipher.SHA256) error {
+	uxouts := uxOut{tx.Bucket(up.pool.Name)}
+	meta := unspentMeta{tx.Bucket(up.meta.Name)}
+
+	for _, hash := range hashes {
+		ux, ok, err := uxouts.get(hash)
+		if err != nil {
+			return err
+		}
+
+		if !ok {
+			continue
+		}
+
+		uxHash, err := meta.getXorHash()
+		if err != nil {
+			return err
+		}
+
+		uxHash = uxHash.Xor(ux.SnapshotHash())
+
+		// update uxhash
+		if err = meta.setXorHash(uxHash); err != nil {
```

Comment:
because of the function needs to have a `bolt.Tx` pass in as argument, all this function have `withTx` suffix are run in transaction that might be rollback if any error